File: src/code-samples/langchain/agent-invocation-thread-and-context.py
# :snippet-start: agent-invocation-thread-and-context-py
from dataclasses import dataclass

from langchain.agents import create_agent
from langchain.messages import AIMessage
from langchain_core.language_models.fake_chat_models import GenericFakeChatModel
from langchain_core.utils.uuid import uuid7
from langgraph.checkpoint.memory import InMemorySaver


@dataclass
class Context:
    user_id: str


agent = create_agent(
    model=GenericFakeChatModel(messages=iter([AIMessage(content="ok")])),
    tools=[],
    context_schema=Context,
    checkpointer=InMemorySaver(),
)

result = agent.invoke(
    {"messages": [{"role": "user", "content": "What's the weather in San Francisco?"}]},
    config={"configurable": {"thread_id": str(uuid7())}},
    context=Context(user_id="user-123"),
)
# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    assert result["messages"][-1].content_blocks[0]["text"] == "ok"
    print("✓ thread_id and context invoke together without error")
# :remove-end:

File: src/code-samples/langchain/agent-invocation-thread-and-context.ts
// :snippet-start: agent-invocation-thread-and-context-js
import { v4 as uuidv4 } from "uuid";
import * as z from "zod";
import { createAgent } from "langchain";
import { AIMessage } from "@langchain/core/messages";
import { FakeListChatModel } from "@langchain/core/utils/testing";
import { MemorySaver } from "@langchain/langgraph";

const contextSchema = z.object({
  user_id: z.string(),
});

const model = new FakeListChatModel({ responses: ["ok"] });

const agent = createAgent({
  model,
  tools: [],
  contextSchema,
  checkpointer: new MemorySaver(),
});

const result = await agent.invoke(
  {
    messages: [
      { role: "user", content: "What's the weather in San Francisco?" },
    ],
  },
  {
    configurable: { thread_id: uuidv4() },
    context: { user_id: "user-123" },
  },
);
// :snippet-end:

// :remove-start:
async function main() {
  const last = result.messages[result.messages.length - 1];
  if (!(last instanceof AIMessage)) {
    throw new Error("expected assistant message");
  }
  const blocks = last.contentBlocks ?? [];
  const text = blocks.find((b) => b.type === "text");
  const body =
    text && text.type === "text"
      ? text.text
      : typeof last.content === "string"
        ? last.content
        : "";
  if (body !== "ok") {
    throw new Error(`unexpected reply: ${body}`);
  }
  console.log("✓ thread_id and context invoke together without error");
}

main();
// :remove-end:
